### DISCUSSION TITLE
Python stub

### DISCUSSION THREAD

**Discussion by pca006132** [elalish/manifold#654] [General]
I was trying to use https://github.com/cansik/nanobind-stubgen to generate python stub for autocomplete and static type check. The stub it generates is not perfect and requires some manual updates. Will open a PR later for this.

I wonder what is the best way to keep the stub up to date with the python file. Should we write our own script to validate them?

The stub file also indicates we can have better numpy support for a lot of our functions, which @wrongbad is doing now.

**Comment by wrongbad** [comment]
Sounds like it should run as part of setup.py to autogen on the fly when building the package. What kinds of issues are you seeing in the output? Probably we should fix those in the tool itself or in the nanobind code so it can be fully automated.

**Reply by pca006132** [reply]
Yeah ideally I would want the tool to generate everythings automatically. Here are the things that I see missing, some are probably simple to fix but some might be harder:

1. Does not handle overloaded constructor.
2. Does not annotate static methods.
3. Does not annotate every overloaded method. I think it does not annotate the first one and jedi is not happy about this.
4. Cannot figure out the type for `glm::vec3`, probably because we use a custom caster.
5. If the return type is a tuple, e.g. tuple with 6 floats, it can only infer `tuple` as the return type.
